### PR TITLE
Typed Builder implementation Part 4: Actions & Declarations

### PR DESCRIPTION
**Pull request by Kukovec**
- [x] Tests added for any new code
- [x] Ran `make fmt-fix` (or had formatting run automatically on all files edited)
- [ ] Documentation added for any new functionality
- [ ] Entry added to [UNRELEASED.md](./UNRELEASED.md) for any new functionality

### CODE REVIEW

File: tla-types/src/main/scala/at/forsyte/apalache/tla/typecomp/ScopedBuilder.scala [hunk 1]
```diff
@@ -25,9 +27,114 @@ import at.forsyte.apalache.tla.typecomp.subbuilder.{
  *     not be scope-correct, since x would have to be typed as Int within the scope defined by this \A operator. Thus,
  *     such an expression cannot be constructed by the builder.
  *
+ * Note that the guarantees are void if `useTrustedEx` is used.
```
Comment (thpani) [inline tla-types/src/main/scala/at/forsyte/apalache/tla/typecomp/ScopedBuilder.scala:30]:
```suggestion
 * These guarantees are void if [[useTrustedEx]] is used.
```

File: tla-types/src/main/scala/at/forsyte/apalache/tla/typecomp/ScopedBuilder.scala [hunk 2]
```diff
@@ -25,9 +27,114 @@ import at.forsyte.apalache.tla.typecomp.subbuilder.{
...
-    extends BaseBuilder with BoolBuilder with ArithmeticBuilder with SetBuilder with SeqBuilder
-    with LiteralAndNameBuilder
+    extends BaseBuilder with BoolBuilder with ArithmeticBuilder with SetBuilder with SeqBuilder with ActionBuilder
+    with LiteralAndNameBuilder {
+
+  /*
+  To use the builder outside of testing scenarios, where the expressions aren't written from scratch,
+  it is necessary to be able to mark certain expressions as "trusted", e.g. when transforming and recombining invariants,
+  or parts of Init/Next.
+  While `build` is safe as a unidirectional implicit conversion from TBuilderInstruction to TLaEx,
```
Comment (thpani) [inline tla-types/src/main/scala/at/forsyte/apalache/tla/typecomp/ScopedBuilder.scala:43]:
```suggestion
  While `build` is safe as a unidirectional implicit conversion from TBuilderInstruction to TlaEx,
```

File: tla-types/src/main/scala/at/forsyte/apalache/tla/typecomp/ScopedBuilder.scala [hunk 3]
```diff
@@ -25,9 +27,114 @@ import at.forsyte.apalache.tla.typecomp.subbuilder.{
...
+  the inverse, `useTrustedEx`, needs to be explicit, to stress the fact that it should be used rarely,
+  typically at most once per transformation, on the initial input.
+   */
+  /**
+   * Creates a `TBuilderInstruction` from a precomputed `TlaEx`.
+   *
+   * Should be used sparingly, and only for expressions that have already passed static analysis.
+   *
+   * Voids correctness guarantees.
```
Comment (thpani) [inline tla-types/src/main/scala/at/forsyte/apalache/tla/typecomp/ScopedBuilder.scala:52]:
```suggestion
   * Creates a `TBuilderInstruction` from a precomputed `TlaEx`. Voids correctness guarantees.
   *
   * Use sparingly, and only for expressions that have already passed static analysis.
```

File: tla-types/src/main/scala/at/forsyte/apalache/tla/typecomp/ScopedBuilder.scala [hunk 4]
```diff
@@ -25,9 +27,114 @@ import at.forsyte.apalache.tla.typecomp.subbuilder.{
...
+  /*
+  To use the builder outside of testing scenarios, where the expressions aren't written from scratch,
+  it is necessary to be able to mark certain expressions as "trusted", e.g. when transforming and recombining invariants,
+  or parts of Init/Next.
+  While `build` is safe as a unidirectional implicit conversion from TBuilderInstruction to TLaEx,
+  the inverse, `useTrustedEx`, needs to be explicit, to stress the fact that it should be used rarely,
+  typically at most once per transformation, on the initial input.
+   */
+  /**
+   * Creates a `TBuilderInstruction` from a precomputed `TlaEx`.
```
Comment (thpani) [inline tla-types/src/main/scala/at/forsyte/apalache/tla/typecomp/ScopedBuilder.scala:48]: Can we just include this paragraph in the Scaladoc below?

File: tla-types/src/main/scala/at/forsyte/apalache/tla/typecomp/ScopedBuilder.scala [hunk 5]
```diff
@@ -25,9 +27,114 @@ import at.forsyte.apalache.tla.typecomp.subbuilder.{
...
+
+  /** x' = y */
+  def primeEq(x: TBuilderInstruction, y: TBuilderInstruction): TBuilderInstruction = eql(prime(x), y)
+
+  type TypedParam = (OperParam, TlaType1)
+
+  /** Parameters to TLA operators can be typed as OperT1 exactly at the top-level, and only if arity > 0 */
+  private def isAcceptableParamType(canContainOper: Boolean): TlaType1 => Boolean = {
+    case FunT1(arg, res)         => isAcceptableParamType(false)(arg) && isAcceptableParamType(false)(res)
+    case SetT1(elem)             => isAcceptableParamType(false)(elem)
```
Comment (thpani) [inline tla-types/src/main/scala/at/forsyte/apalache/tla/typecomp/ScopedBuilder.scala:64]: Can we get a functional description as Scaladoc?

The current comment is more explanatory about the implementation, and should probably go under `case OperT1(args, res)` the function body?

File: tla-types/src/main/scala/at/forsyte/apalache/tla/typecomp/ScopedBuilder.scala [hunk 6]
```diff
@@ -25,9 +27,114 @@ import at.forsyte.apalache.tla.typecomp.subbuilder.{
...
+        } yield ()
+      }
+    } yield TlaOperDecl(opName, params.map(_._1).toList, bodyEx)(
+        Typed(
+            OperT1(params.map { _._2 }, bodyEx.typeTag.asTlaType1())
+        )
+    )
+  }
+
+  /** Attempt to infer the parameter types from the scope. Fails if a parameter name is not in scope. */
```
Comment (thpani) [inline tla-types/src/main/scala/at/forsyte/apalache/tla/typecomp/ScopedBuilder.scala:118]:
```suggestion
  /** Infer parameter types from the scope. Fails if a parameter name is not in scope. */
```

File: tla-types/src/main/scala/at/forsyte/apalache/tla/typecomp/unsafe/UnsafeActionBuilder.scala [hunk 7]
```diff
@@ -0,0 +1,33 @@
+package at.forsyte.apalache.tla.typecomp.unsafe
+
+import at.forsyte.apalache.tla.lir.TlaEx
+import at.forsyte.apalache.tla.lir.oper.TlaActionOper
+
+/**
+ * Unsafe builder for base TlaActionOper expressions.
```
Comment (thpani) [inline tla-types/src/main/scala/at/forsyte/apalache/tla/typecomp/unsafe/UnsafeActionBuilder.scala:7]:
```suggestion
 * Type-unsafe builder for base TlaActionOper expressions.
```

File: tla-types/src/main/scala/at/forsyte/apalache/tla/typecomp/subbuilder/ActionBuilder.scala [hunk 8]
```diff
@@ -0,0 +1,32 @@
+package at.forsyte.apalache.tla.typecomp.subbuilder
+
+import at.forsyte.apalache.tla.typecomp.TBuilderInstruction
+import at.forsyte.apalache.tla.typecomp.unsafe.UnsafeActionBuilder
+import at.forsyte.apalache.tla.typecomp.BuilderUtil.binaryFromUnsafe
+
+/**
+ * Builder for TlaActionOper expressions.
```
Comment (thpani) [inline tla-types/src/main/scala/at/forsyte/apalache/tla/typecomp/subbuilder/ActionBuilder.scala:8]:
```suggestion
 * Type-safe for TlaActionOper expressions.
```

**Review by thpani — APPROVED** [review]
Looks good!
I just have a few suggestions about documentation.

File: tla-types/src/main/scala/at/forsyte/apalache/tla/typecomp/ScopedBuilder.scala [hunk 9]
```diff
@@ -25,9 +27,114 @@ import at.forsyte.apalache.tla.typecomp.subbuilder.{
...
+    params.foreach(validateParamType)
+    for {
+      bodyEx <- body
+      // Check param types against their use in body, then mark as bound
+      _ <- params.foldLeft(().point[TBuilderInternalState]) { case (cmp, (OperParam(pName, _), tt)) =>
+        for {
+          _ <- cmp
+          pNameEx <- name(pName, tt) // throws scopeException if param name clashes with param use in body
+          _ <- markAsBound(pNameEx)
+        } yield ()
```
Comment (konnov) [inline tla-types/src/main/scala/at/forsyte/apalache/tla/typecomp/ScopedBuilder.scala:109]: Do you really need a for-yield loop here? This looks to me like a lot of advanced features for a couple of side effects

File: tla-types/src/main/scala/at/forsyte/apalache/tla/typecomp/subbuilder/ActionBuilder.scala (hunk 8 above)
Comment (konnov) [reply tla-types/src/main/scala/at/forsyte/apalache/tla/typecomp/subbuilder/ActionBuilder.scala:8]: Type-safe builder for TlaActionOper :)

**Review by konnov — APPROVED** [review]
LGTM

File: tla-types/src/main/scala/at/forsyte/apalache/tla/typecomp/ScopedBuilder.scala (hunk 9 above)
Comment (Kukovec) [reply tla-types/src/main/scala/at/forsyte/apalache/tla/typecomp/ScopedBuilder.scala:107]: Yes, with one computation you could get away with `flatMap`, but both `name` and `markAsBound` are computations. Plus, for expands to chained `flatMap` internally anyway.

File: tla-types/src/main/scala/at/forsyte/apalache/tla/typecomp/ScopedBuilder.scala (hunk 4 above)
Comment (Kukovec) [reply tla-types/src/main/scala/at/forsyte/apalache/tla/typecomp/ScopedBuilder.scala]: We could, but the purpose of splitting docs was to keep the Scaladoc lean and limited to descriptions, whereas the normal doc would offer an explanation to devs of why/how the decision was made, which someone using the builder needn't care about.

File: tla-types/src/main/scala/at/forsyte/apalache/tla/typecomp/subbuilder/ActionBuilder.scala (hunk 8 above)
Comment (Kukovec) [reply tla-types/src/main/scala/at/forsyte/apalache/tla/typecomp/subbuilder/ActionBuilder.scala:8]: We have this same doc convention ("Builder" vs "Unsafe builder") in all the other sub-builders, why should this one be different?

File: tla-types/src/main/scala/at/forsyte/apalache/tla/typecomp/ScopedBuilder.scala [hunk 10]
```diff
@@ -25,9 +27,114 @@ import at.forsyte.apalache.tla.typecomp.subbuilder.{
...
+   */
+  def useTrustedEx(ex: TlaEx): TBuilderInstruction = ex.point[TBuilderInternalState]
+
+  /** x' = y */
+  def primeEq(x: TBuilderInstruction, y: TBuilderInstruction): TBuilderInstruction = eql(prime(x), y)
+
+  type TypedParam = (OperParam, TlaType1)
+
+  /** Parameters to TLA operators can be typed as OperT1 exactly at the top-level, and only if arity > 0 */
+  private def isAcceptableParamType(canContainOper: Boolean): TlaType1 => Boolean = {
```
[the remark this answers is not in the document]
Comment (Kukovec) [reply tla-types/src/main/scala/at/forsyte/apalache/tla/typecomp/ScopedBuilder.scala:60]: Sure, I typically write them in dependency order, regardless of private/protected/public status, but I'm not too attached to that.

File: tla-types/src/main/scala/at/forsyte/apalache/tla/typecomp/ScopedBuilder.scala (hunk 9 above)
Comment (Kukovec) [reply tla-types/src/main/scala/at/forsyte/apalache/tla/typecomp/ScopedBuilder.scala:107]: Also, `for-yield` is hardly an "advanced feature".

File: tla-types/src/main/scala/at/forsyte/apalache/tla/typecomp/ScopedBuilder.scala (hunk 4 above)
Comment (thpani) [reply tla-types/src/main/scala/at/forsyte/apalache/tla/typecomp/ScopedBuilder.scala]: Why do we need the Scaladoc to be lean and limited?
It's [quite common](http://hg.openjdk.java.net/jdk8/jdk8/jdk/file/f92ab6dbbff8/src/share/classes/java/lang/Long.java#l465) to have longer doc comments if necessary.

> To use the builder outside of testing scenarios [...]

is definitely relevant to someone calling the method.

File: tla-types/src/main/scala/at/forsyte/apalache/tla/typecomp/subbuilder/ActionBuilder.scala (hunk 8 above)
Comment (thpani) [reply tla-types/src/main/scala/at/forsyte/apalache/tla/typecomp/subbuilder/ActionBuilder.scala:8]: Then I overlooked it on the other ones.
I think it's better to be specific.

File: tla-types/src/main/scala/at/forsyte/apalache/tla/typecomp/ScopedBuilder.scala (hunk 10 above)
[the remark this answers is not in the document]
Comment (thpani) [reply tla-types/src/main/scala/at/forsyte/apalache/tla/typecomp/ScopedBuilder.scala:60]: I'd keep the current grouping, it makes more sense if you read the code top to bottom.
Just my 2¢.

File: tla-types/src/main/scala/at/forsyte/apalache/tla/typecomp/ScopedBuilder.scala (hunk 4 above)
Comment (konnov) [reply tla-types/src/main/scala/at/forsyte/apalache/tla/typecomp/ScopedBuilder.scala]: I also do not get the point of hiding the docs in a place where but a few people can see it. You could mark the paragraph as a detailed description.

**Pull request merged**